rclcpp/include/rclcpp/executor.hpp

```diff
@@ -350,6 +351,51 @@ class Executor
   virtual void
   spin_once(std::chrono::nanoseconds timeout = std::chrono::nanoseconds(-1));
 
+  /// Spin (blocking) until the condition is complete, it times out waiting,
+  /// or rclcpp is interrupted.
+  /**
+   * \param[in] future The condition which can be callable or future type to wait on.
+   *   If this function returns SUCCESS, the future can be
+   *   accessed without blocking (though it may still throw an exception).
+   * \param[in] timeout Optional timeout parameter, which gets passed to Executor::spin_node_once.
+   *   `-1` is block forever, `0` is non-blocking.
+   *   If the time spent inside the blocking loop exceeds this timeout, return a TIMEOUT return
+   *   code.
+   * \return The return code, one of `SUCCESS`, `INTERRUPTED`, or `TIMEOUT`.
+   */
+  template<typename ConditionT, typename DurationT = std::chrono::milliseconds>
+  FutureReturnCode
+  spin_until_complete(
+    const ConditionT & condition,
+    DurationT timeout = DurationT(-1))
+  {
+    if constexpr (std::is_invocable_v<ConditionT>) {
+      using RetT = std::invoke_result_t<ConditionT>;
+      static_assert(
+        std::is_same_v<bool, RetT>,
+        "Conditional callable has to return boolean type");
+      return spin_until_complete_impl(condition, timeout);
+    } else {
+      auto check_future = [&condition]() {
+          return condition.wait_for(std::chrono::seconds(0)) ==
+                 std::future_status::ready;
+        };
+      return spin_until_complete_impl(check_future, timeout);
+    }
```

Comment:
I know that this was discussed last time we tried to merge all of these prs (last release), and I messed that up somewhat due to merging it incorrectly (sorry about that), but since we're here again, I do want to (re)ask a fundamental question (https://github.com/ros2/rclcpp/pull/1874#discussion_r835586538) here, do we need to deprecate the `spin_until_future_complete()` API and have this logic in the new API?

Just looking at it again now, it seems like the pros and cons of deprecating it are:

- Pro: one fewer methods to document and test
- Con: many downstream changes to only the function name being called
- Con: complicates the interface of the new `spin_until_complete()` by making it take various types, and therefore requires you read the docstring to know what the `ConditionT` can be
- Con: uses template deduction when overloading could work (e.g. two `spin_until_complete()` methods that take different args), which is harder for text editors and other tools to handle

Am I missing anything?

If not, this seems to be a poor tradeoff and I'd argue that keeping the syntactic sugar function `spin_until_future_complete()` and having it just call the new `spin_until_complete()` is far simpler than deprecating it and complicating this function.

@hliberacki @SteveMacenski @clalancette @ivanpauno @alsora @fujitatomoya CC you all since you were all part of the original discussion a year ago.

Reply:
> Con: complicates the interface of the new spin_until_complete() by making it take various types, and therefore requires you read the docstring to know what the ConditionT can be

At least it can then take various types :smile: 

> Con: uses template deduction when overloading could work (e.g. two spin_until_complete() methods that take different args), which is harder for text editors and other tools to handle

I'd be OK with that. I just want the spin APIs to be independent of specifically futures to allow for spinning other types of things in the future -- for example with a simpler `SimpleServiceClient`-esk object to spin the object which itself contains some future information internally that a user shouldn't have to worry about for most normal cases.  

If that's done via overloading or templating, I'm fine with either. The main objective I had was renaming the method to be more general and take up less space to specify what its arguments tell you is the case: `spin_until_future_complete(future, timeout)` is less natural than `spin_until_complete(future, timeout)`

Reply:
I do agree that the type of `condition` is confusing.

I personally do not have a strong opinion; I was just trying to finally take care of this big set of PRs.

Reply:
> keeping the syntactic sugar function spin_until_future_complete() and having it just call the new spin_until_complete() is far simpler than deprecating it and complicating this function.

actually this is reasonable, i am inclined to be away from deprecating. they can live together.

Reply:
> I'd be OK with that. I just want the spin APIs to be independent of specifically futures to allow for spinning other types of things in the future -- for example with a simpler `SimpleServiceClient`-esk object to spin the object which itself contains some future information internally that a user shouldn't have to worry about for most normal cases.

Let me ask a question; does this PR implement the functionality to spin on "something else" other than a future?

If it does, then I suggest what we do is to modify this PR to reimplement `spin_until_future_complete` in terms of that function.

If it doesn't, then I suggest we just close this.  Without a second user, we are almost certain to get the API wrong and then have to add another variant.

In either case, I think we should *not* deprecate or remove `spin_until_future_complete`.  That is well-named (even if narrow), and there is no reason to cause tons of work downstream to fix that.  That also means that I think we should close all of the downstream PRs from this, since everything downstream will still compile and work as before.

Reply:
> spin_until_future_complete(future, timeout) is less natural than spin_until_complete(future, timeout)

Yeah, I guess I don't agree on that one, but it is subjective. But either way, it's pretty disruptive to change it to be a single name. And I have a hard time arguing (to myself) that it's inaccurately named or that it's misleading or something like that, which would be a more compelling reason to change it despite the disruption.

So, I'd prefer to avoid the rename myself.

> Let me ask a question; does this PR implement the functionality to spin on "something else" other than a future?

Yes, it just expects a callable object and calls it each time the executor wakes up, returning if the callable returns true or it timesout, otherwise it continues to spin.

Reply:
Alright, in conclusion:

1. Do add `spin_until_complete`, but with a callable, _not_ templated to accept a callable or a future
1. Do not deprecate `spin_until_future_complete`
1. Do implement `spin_until_future_complete` using `spin_until_complete`
1. Do the same for `rclpy`
1. Do close downstream PRs; no need to rename calls in downstream packages from `spin_until_future_complete` to `spin_until_complete`

Can you confirm?

Reply:
That seems like a good plan to me.